lib/SILGen/SILGenPoly.cpp

```diff
@@ -3407,3 +3407,113 @@ void SILGenFunction::emitProtocolWitness(Type selfType,
   scope.pop();
   B.createReturn(loc, reqtResultValue);
 }
+
+//===----------------------------------------------------------------------===//
+// Conversion to Canonical SILFunctionType Thunks
+//===----------------------------------------------------------------------===//
+
+static void buildCanonicalFunctionThunkBody(SILGenFunction &SGF,
+                                            SILLocation loc,
+                                            CanSILFunctionType nonCanonicalTy,
+                                            CanSILFunctionType canonicalTy) {
+  SGF.F.setBare(IsBare);
+  SGF.F.setThunk(IsThunk);
+
+  FullExpr scope(SGF.Cleanups, CleanupLocation::get(loc));
+  FormalEvaluationScope formalEvalScope(SGF);
+
+  // Collect the thunk params, creating arguments for each parameter.
+  SmallVector<ManagedValue, 8> origParams;
+  SGF.collectThunkParams(loc, origParams);
+
+  AbstractionPattern canonicalTyPattern(canonicalTy);
+  AbstractionPattern nonCanonicalTyPattern(nonCanonicalTy);
+
+  // Then translate our parameters into new params.
+  SmallVector<ManagedValue, 8> newParams;
+  TranslateArguments(SGF, loc, llvm::makeArrayRef(origParams).drop_back(1),
+                     newParams, nonCanonicalTy->getParameters())
+      .translate(canonicalTyPattern, canonicalTy, nonCanonicalTyPattern,
+                 nonCanonicalTy);
+
+  // Then grab the function we are going to call from the last parameter.
+  ManagedValue fn = origParams.back();
+
+  // Collect the arguments.
+  SmallVector<SILValue, 8> args;
+
+  // Add all of the indirect results. We can just add the SILValues directly to
+  // the args array since we do not need to perform any transformations upon
+  // them because:
+  //
+  // 1. Reabstraction can not occur as a result of a canonical/non-canonical
+  // mismatch.
+  // 2. SILGenFunction::collectThunkParams(...) does not create cleanups when it
+  // creates arguments.
+  SILFunctionConventions fnConv(canonicalTy, SGF.SGM.M);
+  transform(range(fnConv.getNumIndirectSILResults()), std::back_inserter(args),
```

Comment:
Where is transform() defined?

Reply:
That is just std::transform. We define it in STLExtras.

Reply:
I see, I was confused because there are a bunch of methods named transform() in the same source file. Maybe qualify it with std:: to avoid this confusion.

Reply:
no it is a swift wrapper function that applies std::transform top a range. What if I add a nice comment here?